idxmgmt/supporter_factory.go

```diff
@@ -153,24 +155,21 @@ func (cfg *IndexManagementConfig) logWarnings(log *logp.Logger) {
 	}
 }
 
-// unpackTemplateConfig merges APM-specific template settings with (possibly nil)
-// user-defined config, unpacks it over template.DefaultConfig(), returning the result.
-func unpackTemplateConfig(userTemplateConfig *common.Config) (template.TemplateConfig, error) {
-	templateConfig := common.MustNewConfigFrom(`
-settings:
+func mergeDefaultConfig(configRoot *common.Config) (*common.Config, error) {
+	defaultConfig := common.MustNewConfigFrom(`
+setup.template.settings:
   index:
     codec: best_compression
     mapping.total_fields.limit: 2000
     number_of_shards: 1
   _source.enabled: true`)
-	if userTemplateConfig != nil {
-		if err := templateConfig.Merge(userTemplateConfig); err != nil {
-			return template.TemplateConfig{}, errors.Wrap(err, "merging failed")
-		}
-	}
-	out := template.DefaultConfig()
-	if err := templateConfig.Unpack(&out); err != nil {
-		return template.TemplateConfig{}, err
-	}
-	return out, nil
+	if configRoot == nil {
+		return defaultConfig, nil
+	}
+	// NOTE(axw) it's important that we merge onto the root config,
+	// due to how config variable resolution works; variables are
+	// resolved using the root of the left-most config in the merge.
+	// We merge the user-defined config back over the defaults to
+	// ensure they take precedence.
+	return common.MergeConfigs(configRoot, defaultConfig, configRoot)
```

Comment:
This looks complicated - looking at the `common.MergeConfigs` code I couldn't figure out why we need the extra `configRoot`.

Reply:
Agreed, hence the NOTE :)
Is the comment unclear?

Reply:
I tweaked the comment slightly, hopefully it's a bit clearer now.